Repeaters, v.3: Only query when a repeater is unlocked

kaapstorm opened the pull request:
## Technical Summary

This past weekend I was (over?)thinking about how PR #35033 loops over repeaters, and I came up with two alternatives. This "version 3" is the second. (See PR #35686 for version 2.)

### Comparison

|                                 | PR #35033 | Version 2 | Version 3        |
| ------------------------------- | --------- | --------- | ---------------- |
| Finds new repeat records every  | 5 minutes | 1 minute  | 1 minute         |
| Queries for all repeaters every | 5 minutes | Up to every 200ms | As soon as needed |
| Queries per minute              | N/A       | ~60/min   | ~120/min         |
| Rank in order of simplicity     | 1         | 2         | 3                |
| Rank in order of throughput     | 3         | 2         | 1                |

The approach used in PR #35033 is relatively simple, and fast at sending repeat records of repeaters that are already being processed. But new repeat records of repeaters that are not being processed can take up to 5 minutes to be discovered and processed.

In comparison, this version, version 3, checks every minute for repeaters with repeat records ready to be sent (and if it is busy processing repeaters, it will find new repeat records sooner than that).

Instead of waiting and trying again when all repeaters are locked, like version 2, version 3 intelligently queries as soon as a repeater is unlocked. A Redis lock's `acquire(blocking=True)` method polls Redis every 100ms, so version 3 uses the same approach to block while repeaters are processing, and unblocks as soon as the first one is unlocked. This ensures that every query returns a resultset that has at least one repeater that can be processed.

From the "Iterated all repeaters once" chart in Datadog below, you can see that this approach is most effective when there are only a couple of repeaters with repeat records ready to be sent. Version 3 has the highest throughput of all the approaches.

![image](https://github.com/user-attachments/assets/b380359c-7a08-4964-a22b-53e7dace5b7b)

Review commits together :office: 

## Feature Flag

`PROCESS_REPEATERS`

## Safety Assurance

### Safety story

* Tested on Staging
* Not yet in use on Production

### Automated test coverage

Tests included

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

kaapstorm committed:
Start every minute, sleep 0.2s if all locked

kaapstorm committed:
Wait for first process_repeater() to finish

mkangia committed:
merge master

kaapstorm committed:
Tweak merge

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -316,8 +316,26 @@ def process_repeaters():
-            process_repeater(repeater_id, lock_token)
+        group = 0
+        client = get_redis_client()
+        client.set(f'repeater_group_{group}', 0)
```
I think we want this key to expire at some point because we don't want to leave old/unused keys in redis indefinitely.

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -316,8 +316,26 @@ def process_repeaters():
+                        # Wait until (at least) the first task finishes
+                        # so that the next query for repeaters returns a
+                        # new resultset.
+                        time.sleep(0.1)
```
It looks like you're trying to reinvent a blocking lock here, and seems inefficient to be waiting in a loop like this. I haven't thought hard enough about other approaches to make a suggestion for how to make this better, but this does not seem like a good approach.

kaapstorm replied:
This is a blocking lock. It blocks the same way that redis-py's [Lock.acquire()](https://github.com/redis/redis-py/blob/e0906519/redis/lock.py#L209-L218) blocks. It waits for _any_ task currently processing a repeater from the most recent resultset from `Repeater.objects.get_all_ready_ids_by_domain()` to complete. It does this to ensure that the next call to `Repeater.objects.get_all_ready_ids_by_domain()` will return an updated resultset.

The difference between what's currently in `master` and this PR, is that the current code is good at processing the repeat records of the repeaters first returned by `Repeater.objects.get_all_ready_ids_by_domain()`, but repeat records for different repeaters have to wait up to 5 minutes. The approach in this PR will pick them up the next time `Repeater.objects.get_all_ready_ids_by_domain()` is called.

In the old approach, new repeat records are first attempted immediately, so 5 minutes feels like a long time to wait in comparison.

(It could be that this PR is massively over-engineered. A much simpler compromise could be just to set `PROCESS_REPEATERS_INTERVAL = timedelta(minutes=1)`. New repeat records will always have to wait, but at least they will only wait up to a minute.)

millerdev commented on corehq/motech/repeaters/tasks.py:
```diff
@@ -329,22 +347,16 @@ def iter_repeater_id_tokens():
+        metrics_counter('commcare.repeaters.process_repeaters.iter_once')
+        for repeater_id in repeater_ids:
+            lock = RepeaterLock(repeater_id)
+            if lock.acquire():
```
This will block until the lock is acquired. Is the `if` needed?

millerdev reviewed (commented):
I'm going to need more time to evaluate how this new algorithm works. A pair review may be helpful.

kaapstorm committed:
iter_repeater_id_tokens() into process_repeaters()

`iter_repeater_id_tokens()` just made things more complicated than they
need to be. Moving it up into `process_repeaters()` avoids the need to
count repeaters.

Also adds much-needed tests for `process_repeaters()`.

kaapstorm committed:
Use the correct Redis client

MartinRiese committed:
Merge branch 'master' into nh/iter_repeaters_3

MartinRiese commented:
Did not look like a real conflict. Just some changes in the same file adding constants. kept both

kaapstorm committed:
Fix tests, and stupid mistake

kaapstorm committed:
Don't block if no repeaters were acquired

kaapstorm commented:
I'm not sure that the increase in performance makes up for the increase in complexity. Closing.
